bin/node/runtime/src/lib.rs

```diff
@@ -714,6 +715,12 @@ parameter_types! {
 	pub const MaxDepth: u32 = 32;
 	pub const StorageSizeOffset: u32 = 8;
 	pub const MaxValueSize: u32 = 16 * 1024;
+	pub DeletionWeightLimit: Weight = Perbill::from_percent(20) *
+		RuntimeBlockWeights::get().max_block;
```

Comment:
Having multiple `on_initialize` functions that want to take some non-substantial amounts of weight from the block are a problem. They need to be coordinated somehow. Moving the lazy removal from `on_initialize` to some potential [on_idle](https://github.com/paritytech/substrate/issues/4064) could be a solution.

Reply:
That said, I am open to change those numbers here to whatever works.

Reply:
I guess one way could be to have `DeletionWeightLimitRatio < AVERAGE_ON_INITIALIZE_RATIO`. so that even if the chain is busy removing lots of tries, election is still fine.

But I agree `on_idle` could be helpful, maybe staking could notify `on_idle` that it needs some space somehow

Reply:
> I guess one way could be to have DeletionWeightLimitRatio < AVERAGE_ON_INITIALIZE_RATIO. so that even if the chain is busy removing lots of tries, election is still fine.

But this assumes that this would be the only `on_initialize` in the block which it isn't.  The `AVERAGE_ON_INITIALIZE_RATIO` shared between all of them.

Don't we already have this problem because of this code already:
https://github.com/paritytech/substrate/blob/0fc832988bef7ab27f6cf2f3040fcb2d3c9b6568/bin/node/runtime/src/lib.rs#L818-L821

Reply:
>  But this assumes that this would be the only on_initialize in the block which it isn't. The AVERAGE_ON_INITIALIZE_RATIO shared between all of them.

Currently the constant cost on_initialize is very low, just some pallet have high usage but in some blocks, although maybe offences can take multiple block I thought it shouldn't take too many blocks.

I was worry that a big contract removal would take quite long to remove and would almost always trigger the on-chain election. But after all I actually don't know precisely how many block it would take to remove them, and how much congestion will happen there.

Maybe it is just fine as it is, and contract removal shouldn't take too much blocks ?

Reply:
If we allow it to take 20% of the block we can expect it to use it all the time when contracts are heavily used. You cannot really tell, though.  You can expect that it is able to remove around `3000` trie keys in that 20% allotment. If we are removing contracts at that rate or faster it will always take these 20%. We could tune it down to 10% (which is the current `AVERAGE_ON_INITIALIZE_RATIO`) to be on the safe side.

Reply:
I tuned it down to `AVERAGE_ON_INITIALIZE_RATIO`. With that setting we get the following throughput of the lazy removal:
```
Lazy deletion throughput per block (empty queue, full queue): 1954, 1564
```

In order to increase this throughput in the future we should look into `on_idle` or off chain workers.

Reply:
The current `OffchainSolutionWeightLimit` is a bit tricky because it essential to the system to work properly, yet it is an unsigned transaction, so it is always likely that sum of `on_initialize` causes it to fail, in which case we do the notorious on-chain election. This will be fixed with #7319 which is far from being shipped yet, but almost ready. Once this is merged, we do hope to do most of the solution processing in `on_initialize` as well. In which case, the competition becomes fair. Something like staking, because it is critical to the system, would stay in `on_initialize`, and this can move to lazy removal. 

A little brain dump about that: Let's say that a block has 100 weight units. Usually, 10 is consumed in on_initialize, and 40 in extrinsic, leaving 50 unused. Spreading this into the on_idle cases resembleas a thread/process scheduling situation. Two important notes come to mind: 

1. Can we enforce the weight of each `on_idle {}` call? No, the runtime should have trust within it and the on_idle call of a module should simply return the weight that it has consumed, similar to `on_initialize`. 
2. How can we ensure that the idle-scheduling is fair? This can be tricky but I guess a number of policies can be used, and ideally we can have it configurable. Naively, we can start by running the on_idle of each pallet until there is no weight left. Else, we can do it in a ring fashion, where if pallet at indices `{0..3}` got to execute their on_idle on this block, on the next block we start from `{4..}`.

Reply:
Shouldn't we discuss this in the `on_idle` issue? Or do you think the lazy deletion would essentially break elections and therefore we need to implement `on_idle` first?

Reply:
I added the following issue to keep track of this issue: #7817

That said, I will merge in the current form because:
* This configuration is only for the dev node and neither kusama nor polkadot have it deployed
* This is only an issue when contracts are heavily utilized the time an offchain election happens
* Having an on-chain election will not break the chain
* #7319 will most likely be merged before it gets important
* on_idle will most likely be available before it gets important

Reply:
I don't have any objections with merging this as it is. Thanks for making the issue!